source/comp/markv_codec.cpp

```diff
@@ -0,0 +1,1194 @@
+      // have no type Id, and will map to 0. The result Id for a
+      // type-generating instruction (e.g. OpTypeInt) maps to itself.
+      auto insertion_result = id_to_type_id_.emplace(
+          inst->result_id,
+          spvOpcodeGeneratesType(opcode) ? inst->result_id : inst->type_id);
+      if(!insertion_result.second) {
+        return vstate_.diag(SPV_ERROR_INVALID_ID)
+            << "Unexpected behavior: id->type_id pair was already registered";
+      }
+      break;
+    }
+
+    case SPV_OPERAND_TYPE_ID:
+    case SPV_OPERAND_TYPE_OPTIONAL_ID:
+    case SPV_OPERAND_TYPE_SCOPE_ID:
+    case SPV_OPERAND_TYPE_MEMORY_SEMANTICS_ID: {
+      uint32_t id = 0;
+      if (!DecodeOperandWord(type, &id))
+        return vstate_.diag(SPV_ERROR_INVALID_BINARY) << "Failed to read id";
+
+      if (id == 0)
+        return vstate_.diag(SPV_ERROR_INVALID_BINARY) << "Decoded id is 0";
+
+      spirv_.push_back(id);
+      vstate_.setIdBound(std::max(vstate_.getIdBound(), id + 1));
+
+      if (type == SPV_OPERAND_TYPE_ID || type == SPV_OPERAND_TYPE_OPTIONAL_ID) {
+
+        parsed_operand.type = SPV_OPERAND_TYPE_ID;
+
+        if (opcode == SpvOpExtInst && parsed_operand.offset == 3) {
+          assert(0);
+        }
+      }
+      break;
+    }
+
+    case SPV_OPERAND_TYPE_EXTENSION_INSTRUCTION_NUMBER: {
+      assert(0);
+      break;
+    }
+
+    case SPV_OPERAND_TYPE_LITERAL_INTEGER:
+    case SPV_OPERAND_TYPE_OPTIONAL_LITERAL_INTEGER: {
+      // These are regular single-word literal integer operands.
+      // Post-parsing validation should check the range of the parsed value.
+      parsed_operand.type = SPV_OPERAND_TYPE_LITERAL_INTEGER;
+      // It turns out they are always unsigned integers!
+      parsed_operand.number_kind = SPV_NUMBER_UNSIGNED_INT;
+      parsed_operand.number_bit_width = 32;
+
+      uint32_t word = 0;
+      if (!DecodeOperandWord(type, &word))
+        return vstate_.diag(SPV_ERROR_INVALID_BINARY)
+            << "Failed to read literal integer";
+
+      spirv_.push_back(word);
+      break;
+    }
+
+    case SPV_OPERAND_TYPE_TYPED_LITERAL_NUMBER:
+    case SPV_OPERAND_TYPE_OPTIONAL_TYPED_LITERAL_INTEGER:
+      parsed_operand.type = SPV_OPERAND_TYPE_TYPED_LITERAL_NUMBER;
+      if (opcode == SpvOpSwitch) {
+        assert(0);
+      } else {
+        assert(opcode == SpvOpConstant || opcode == SpvOpSpecConstant);
+        // The literal number type is determined by the type Id for the
+        // constant.
+        assert(inst->type_id);
+        if (auto error =
+            SetNumericTypeInfoForType(&parsed_operand, inst->type_id))
+          return error;
+
+        if (auto error = DecodeLiteralNumber(parsed_operand))
+          return error;
+      }
+      break;
+
+    case SPV_OPERAND_TYPE_LITERAL_STRING:
+    case SPV_OPERAND_TYPE_OPTIONAL_LITERAL_STRING: {
+      parsed_operand.type = SPV_OPERAND_TYPE_LITERAL_STRING;
+      std::vector<char> str;
+      while (true) {
```

Comment:
I looked further. The reader_ is your own creation rather than a reader passed in.  So you are relying on the boundedness of the memory buffer, correctness of the reader code.  If the input buffer's string is not null-terminated, then I presume the ReadUnencoded method will return false and you'll get the invalid binary.

Probably worth a comment, since infinite loops are so easily dangerous.